src/AutoRest.CSharp.V3/Output/Builders/SerializationBuilder.cs

```diff
@@ -150,26 +153,82 @@ private static XmlObjectSerialization BuildXmlObjectSerialization(ObjectSchema o
+            property.SerializedName,
+            property.CSharpName(),
+            BuildSerialization(property.Schema, property.IsNullable(), property.IsBag),
+            property.IsBag
+        );
+
         private static JsonObjectSerialization BuildJsonObjectSerialization(ObjectSchema objectSchema, TypeReference schemaTypeReference, bool isNullable)
         {
             List<JsonPropertySerialization> serializationProperties = new List<JsonPropertySerialization>();
             foreach (var schema in EnumerateHierarchy(objectSchema))
             {
-                foreach (Property property in schema.Properties!)
+                foreach (Property property in AggregateProperties(schema))
                 {
-                    JsonPropertySerialization propertySerialization = new JsonPropertySerialization(
-                        property.SerializedName,
-                        property.CSharpName(),
-                        BuildSerialization(property.Schema, property.IsNullable())
-                    );
-
-                    serializationProperties.Add(propertySerialization);
+                    serializationProperties.Add(CreateJsonPropertySerialization(property));
                 }
             }
 
             return new JsonObjectSerialization(schemaTypeReference, serializationProperties.ToArray(), CreateAdditionalProperties(objectSchema));
         }
 
+        private class PropertyBag
+        {
+            public Dictionary<string, PropertyBag> Bag { get; } = new Dictionary<string, PropertyBag>();
+            public List<Property> Properties { get; } = new List<Property>();
+        }
+
+        private static IEnumerable<Property> AggregateProperties(ObjectSchema schema)
+        {
+            PropertyBag propertyBag = new PropertyBag();
+            propertyBag.Properties.AddRange(schema.Properties ?? Array.Empty<Property>());
+            PopulatePropertyBag(propertyBag, 0);
+            return propertyBag.Properties;
+        }
+
+        private static Property CreatePropertyFromPropertyBag(string name, PropertyBag bag) => new Property
+        {
+            IsBag = true,
+            SerializedName = name,
+            Schema = new ObjectSchema
+            {
+                Properties = bag.Properties,
+                Language = new Languages { Default = new Language { Name = $"SYNTHETIC-{name}" } },
+                Type = AllSchemaTypes.Object
+            },
+            Required = true,
+            Language = new Languages { Default = new Language { Name = name } }
+        };
+
+        private static void PopulatePropertyBag(PropertyBag propertyBag, int depthIndex)
+        {
+            foreach (Property property in propertyBag.Properties.ToArray())
+            {
+                if (!(property.FlattenedNames?.Any() ?? false) || depthIndex >= property.FlattenedNames!.Count - 1)
```

Comment:
this condition is hella complicated, simplify it

Reply:
Done.